---
title: "simplex: An individual-based platform for ecological modeling"
author: "Kenneth J. Locey"
fontsize: 12pt
geometry: margin=2.54cm
output: pdf_document
header-includes:
    - \usepackage{setspace}
    - \doublespacing
---

\begin{center} 
Department of Biology, University of Indiana, Bloomington
\end{center}

\begin{center}
email: kjlocey@indiana.edu
\end{center}

\pagebreak

**ABSTRACT**

A popular approach in ecological modeling explicitly simulates the biology and ecology of individual organisms. Individual-based models (IBM) allow ecological patterns and dynamics to be studied as they emerge from the individual level according to simulated principles and theories. IBMs also allow virtually any aspect of a simulated system to be tracked and recorded for later analysis. I developed a novel IBM platform (i.e, simplex) to explicitly simulate physiology, life history, resource-limited growth, fluctuating environmental dynamics, and community assembly and structure in both fluid and non-fluid environments. simplex accomplishes three primary tasks in keeping with its namesake (i.e., a generalized notion of a triangle to arbitrary dimensions). First, simplex assembles and runs stochastic ecological models from random combinations of over 20 state variables. Second, simplex stores the output of models and generates and stores animations. Third, simplex provides R Markdown files for analyzing simulated data. The simplex source code is freely available, includes the standard ODD description protocol, is open to community-based revision, includes unit testing and time testing, and has been developed to run on modest hardware.

\pagebreak

**INTRODUCTION**

\setlength\parindent{36pt}

Modeling is a elementary approach to understanding ecological systems, the influence of ecological processes, and the predictability of ecological patterns and dynamics. Before modern computing, ecological models were almost exclusively equation-based representations of highly simplified systems (Black and McKane 2012, Otto and Troy 2007, Grant and Swannack 2008). Since the advent of personal computing, ecological models have been increasingly constructed to handle greater complexity and to explicitly simulate ecological processes.

In ecology, simulation-based models are often used to examine analytically challenging or intractable scenarios. 
Examples are markov models that simulate stochastic demographic changes and early ecological null models that , both of which operate on some form of matrix (Gotelli and Entsminger 2001, Hubbell 2001). Ecologists have also simulated growth and interactions among individuals for over two decades with individual-based models (IBMs) (DeAngelis and Gross 1992, Rosindell et al. 2015). 
IBMs explicitly encode rules of how individuals change and interact. Once the rules are encoded, population to ecosystem-level dynamics can emerge as an IBM simulates over time and spatially explicit environments.

A body of literature including original research, comprehensive reviews, and textbooks reveal the use, advantages, and challenges of ecological IBMs (e.g., Grimm 1999, Grimm and Railsback 2005). IBMs can provide degrees of ecological realism, individual variability, and spatial heterogenetiy that are unattainable with other models. IBMs also offer the potential for realistic and unanticipated ecological dynamics and patterns to emerge from individual level interactions. However, IBMs come with challenges that include greater computational complexity, the difficulty of explicitly encoding ecological theory, and the challenge of using IBMs to solve applied problems (Matthews et al. 2007, Grimm and Railsback 2005).

Despite their frequent use, the power of ecological IBMs has yet been leveraged to its greatest advantage. Few, if any, include fluid dynamics or combine fluid dynamics with growth and dispersal. While growth, sensing, and decision making are often modeled, few IBMs integrate physiology, nutrient-limited growth, evolution, environmental flux, community ecology, biogeography, and stochastic sampling theory. Ecological IBMs are more often constructed to model a specific system than to synthesize general theories of ecology and evolution (but see Rosindell et al. 2015). Yet, IBMs can allow researchers to creatively and explicitly simulate any number of biological and ecological theories and principles. Likewise, IBMs can made to track and analyze any information from physiological changes to community assembly, and the structure of an entire ecosystem.

Here, I present a detailed description for an ecological IBM platform referred to as "simplex". The simplex platform simulates a broad range of ecological conditions, records information from individuals to the ecosystem level, and provides computing code for analyzing output. simplex incorporates fluid dynamics, stochastic and selection-based changes within and among populations, and integrates nutrient-limited growth and physiology. I constructed simplex with the aim of simulating ecologically complex scenarios, from which the results could be used to test (*i*) predictions of ecological theory and (*ii*) the general influence of constraints and processes on populations, communities, and ecosystems. I provide an explanation of how simplex works, the data it quantifies and tracks, the theories and principles that simplex integrates, and the analyses that can be conducted using the open source code.

\setlength\parindent{0pt}

**METHODS**

**Model description following ODD Protocol**  
The ODD protocol is standard for describing individual-based models (Grimm et al. 2006). ODD stands for Overview, Design concepts, and Details. Here, I provide an abbreviated explanation of the simplex ODD protocol. The full ODD protocol can be found on the public GitHub repository at https://github.com/LennonLab/simplex. Likewise, to save space, I refer the reader to the README.md file on the public simplex GitHub.com repository (https://github.com/LennonLab/simplex/blob/master/README.md) for descriptions of source code files and directories.

**Purpose**  
The purpose of simplex is to explicitly simulate life history and the evolution of traits among individual organisms, the assembly of ecological communities, and the structure and function of ecosystems in a spatially explicit environment under stochastic conditions. The proximate goal of simplex is to generate high degrees of variation in the assembly and structure of populations and communities by assembling many different models from random combinations of state-variables and processes. The ultimate goal of simplex is to provide a simulation-based platform for examining conditions under which processes and constraints have a robust influence on eco-evolutionary dynamics, patterns of biodiversity, and ecosystem processes.

**Entities & their state variables**

*Individual organisms*--Individuals are distinguished by collections of elements within lists. Individuals undergo changes when randomly sampled from lists. Each specific position in the list corresponds to the same individual. For example, in simulating growth, a simplex model chooses an individual from the list of cell quotas (the probability of reproducing is determined by endogenous resources). The first position in this list as well in all other lists of individual attributes corresponds to the same individual. Lists of attributes for individual organisms include the time each individual spends in the system, species ID, individual ID, 2D spatial location, endogenous levels of resources (resource specific cell quotas), individual-level metabolic maintenance cost, individual-level maximum dispersal rate, parental lineage, individual resource use efficiency for each resource.

*Species*--Each species is characterized by the individuals that share a common set of traits, such as maximum growth rate, metabolic maintenance cost. Species information is stored in Python dictionaries. In this way, if simplex requires the species ID of an individual it will access the species ID list where each element corresponds to a specific individual. But, if simplex requires the maximum growth rate for an individual, then it finds the species ID and then uses that to accesses the Python dictionary for maximum specific growth rates of species. The species-level information that are stored include metabolic maintenance cost, maximum dispersal rate, maximum theoretical growth rate, and the resource use efficiency for each resource.

*Resource particles*--Individual resource particles are distinguished by collections of elements within lists. Each specific position in the list corresponds to the same resource particle. The types of information stored about each resource particle include time in the system, particle ID, 2D spatial location, the type of resource particle, the size of the particle.

*Inert tracers particles*--These are objects that move/flow into and through the environment. They only interact with physical barriers. The use of tracers particles allows for attributes of a system that flows or physically turns over to be quantified. Like other individual-level objects in simplex models, tracers are distinguished by collections of elements within lists. Each specific position in the list corresponds to the same resource particle. Information stored about each resource particle includes time spent in the system, 2D spatial location and particle ID.

*Physcial barriers*--These objects are simulated as discrete 2D spatial coordinates that cannot be occupied by any individual entities. The number, size, and location of physical barriers are chosen at random by simplex at the start of each model.

**System level state variables**  
Each run of simplex begins with random choices for the values of width (5 to 100), height (5 to 100), basal flow rate (0.0 to 1.0), the number, size, and location of physical barriers, the number and direction of environmental gradients, rate of stochastic disturbance, rate of fluctuation in basal flow rate, and degrees of fluctuation and syncronized flow of individual particles, ranging from completely in-sync or completely out of sync.


**Spatial and temporal scales**  
\setlength\parindent{36pt}
The two general aspects of scale are grain (i.e., resolution) and extent (e.g. total area). The environment of simplex models is two dimensional and can vary along each axis from 5 to 100 discrete units. This makes for a potential total extent of 25 to 10,000 discrete patches, each with a grain of 1 square unit. All particles move in decimal units, meaning that individual particles can  practically infinite locations within patches and likewise. Extent of time in simplex models refers to residence time, i.e., the average amount of time that individual particles spend in the system. Residence time for individual particles (organisms, resources, tracers) can vary across five orders of magnitude.

Grain is the smallest unit over which change can happen. For example, the original ODD documentation states “One time step represents one year and simulations were run for 100 years. One grid cell represents 1 ha and the model landscape comprised 1,000 x 1,000 ha; i.e., 10,000 square kilometers”. In contrast, the smallest grain achievable by simplex is determined by slowest rate at which individuals can undergo BIDE (birth, immigration, death, emigration) processes. For example, under high residence times, an individual can move across 0.00001% of the x or y axis in one time step. Under low residence times, an individual can move across 10% or greater of the x or y axis in one time step.

\setlength\parindent{0pt}
**Process overview and scheduling**

*Assembly*. The user runs a program that chooses random values for system-level state variables including whether disturbance, immigration, speciation, fluid dynamics, etc. will occur and at what rates.

*Core simulation process*. simplex models begin simulation immediately after assembly from random combinations of state-variables and processes. Instead of operating by definitive time steps (i.e. days, generations), simplex models advance turnover of the environmental matrix according to the initial rate of flow. If the initial rate of flow is 1.0, then the environmental matrix and inert particles would flow 1.0 units of distance. After each iteration of flow, each individual is given the chance to consume, grow, reproduce, starve, die, and to disperse towards resources and environmental optima.

*Duration: A run to mean reversion*. Once assembled, a simplex model simulates ecological processes (birth, death, dispersal, growth, consumption, etc.) until the system reaches a point of mean reversion, i.e., the tendency of a system to reverse a directional change in, say, total abundance. Mean reversion captures whether a system is fluctuating around a mean value. Once 100 generations have occurred within a model, simplex examines whether a point of mean reversion has occured by conducting an Augmented Dickey-Fuller (ADF) Test, which is well-explained here: https://www.quantstart.com/articles/Basics-of-Statistical-Mean-Reversion-Testing. A model is stopped once mean reversion is determined to occurred and, unless the number of desired simulations (i.e. models) has been reached, simplex simply constructs another model from random combinations of state-variables and processes, and then runs it to mean reversion.

*Fluid dynamics*. simplex uses an efficient and powerful method for simulating fluid flow, i.e., a Lattice-Boltzmann Method (LBM) (Succi 2001). An LBM discretizes the environment into a lattice and attaches to each position in the lattice a number of particle densities and velocities for each simulated direction. simplex simulates nine directions of movement possible in a 2D spatial environment (i.e., N, S, E, W, NE, NW, SE, SW, current position).

*Active dispersal*. simplex models allow individuals to move towards their environmental optima. Rather than a single environmental optima resulting from a single environmental gradient, simplex allows environmental optima to occur as intersections among environmental gradients. Hence, individuals potentially have multiple optima resulting from unique and equally optimal intersection of up to 10 environmental gradients.

*Simulated life history*.\setlength\parindent{36pt} simplex models simulate growth, reproduction, and death via weighted random sampling. This simulate the partly probabilistic and partly deterministic nature of environmental filtering and individual-level interactions.

Inflow/Entrance. Resources and individuals enter from any point in the environment. Species identities of inflowing propagules are chosen at random from a log-series distribution, which often approximates the distribution of abundance among species in ecological communities (see Hubbell 2001). Along with a species ID and species-specific maximum rates of resource uptake, active dispersal, resource efficiencies, cell maintenance, and environmental optima, each propagule was given a unique ID and a multi-resource cell quota that represent the state of internal resources. The average of these cell quotas determine the probability of reproduction. 

Dispersal. Individuals are allowed to actively move along environmental gradients (sometimes against the direction of environmenta flow) and towards their optima. A better match to one's environmental optima increases the chance of reproduction and the individual's ability to perform (consume, grow). 

Consumption & growth. Sampled individuals consume resources according to their specific maximum rates of uptake and grow according to specific resource efficiencies and maintenance costs. Uptake increases the individual cell quotas and decreases ambient resources. Individual cell quotas are then decreased according to a specific physiological maintenance cost. 

Reproduction. Reproduction in simplex is currently limited to clonal reproduction with the possibility of mutation. Individuals reproduce with a probability determined by the combination of mean cell quota and the proportional match to the environmental optima. The cell quota of each resource is evenly divided between the individual and its daughter. The daughter is given a unique individual ID and the species ID of its mother, unless in the case of speciation, but is allowed small mutations in individual-level state variables.

Speciation. Speciation is simulated within simplex as a discrete event (*sensu* Hubbell 2001 and Bell 2001) and is accompanied by mutations in the values of species-level state variables, much like that expected from a Wright-Fisher model (Hartl and Clark 1997). These random events allows for diversity to arise within the system, which the environmental filter can then select on.

Death and emmigration. Individuals sampled at random will die if their smallest resource specific cell quota (i.e., N, C, P) is equal to or less than 0. Individuals, resource particles, and inert tracers are considered to have left or to have flowed out when they pass beyond edges of the environment.

\setlength\parindent{0pt}

**Design concepts**  
*Ecological complexity*. simplex assembles models from random combinations of constraints (state-variables) and processes to generate output data that allow the user to test the general influence of a particular state-variable, process, or combination thereof using univariate and multivariate tests.  

*Nutrient limited growth*. All models assembled by simplex employ the universal concept that individual growth and activity is fueled and limited by resources.  

*Resource diversity & heterogeneity*. simplex allows the user to explore the influence of the number and abundances of different resources on ecological diversity and ecosystem processes.

**Theories**  
*Constraint-based theory*. simplex was originally built to explore the influence of ecosystem residence time (volume/flow rate) on community assembly and structure. That is, the idea that both ecological processes and physical constraints shape ecological diversity. 

*Chemostat theory*. simplex operates much like an unhinged bioreactor or chemostat. That is, particles flow through a system of a defined size at an average rate, and are limited in growth by their residence time.

*Ecological neutral theory*. simplex operates via random sampling and can vary from being completely neutral (all individuals having equal vital rates) to completely idiosyncratic (all individual and species are as different as possible). The one aspect of neutral theory that simplex adopts without question is the importance of stochastic life history processes (i.e. weighted or unweighted random fluctuations in population sizes). 

**Adaptation**
Individuals can move towards their environmental optima. Populations can become aggregated in areas that provide favorable intersections of species optima. Species can evolve by the action of the environmental filter on subpopulation variation in state variables. Individual can seek conditions that match them best to their environment.

**Sensing**  
Individuals can move towards environmental optima against the flow of the environment. Otherwise, all encounters are the result of random walks or fluid flow.

**Interaction**  
At the moment, individuals only interact indirectly through excluding each other from resources (e.g. preemption).

**Stochasticity**  
The occurrence of nearly all processes of birth, death, life, immigration, dispersal, emigration, consumption, etc. are conducted via random sampling. In this way, population and community dynamics result, in part, from demographic 
stochasticity. Likewise, the emergence of life history traits proceeds from initially random combinations of traits.

**Observation**  
The following is recorded for each simplex model:

* Values of randomly chosen state variables
* Total abundance ($N$) and Species richness ($S$)
* mean growth rate (per species & per capita) and mean maintenance cost (per species & per capita) 
* mean resource efficiency (per species & per capita) and mean active dispersal rate (per species & per capita) 
* Compositional turnover: Bray-Curtis, Sorensen's
* Whittaker's species turnover
* Species evenness: Smith and Wilson's index ($E_{var}$), Simpson's index ($E_{1/D}$) 
* Species diversity: Shannon's diversity ($H'$), Simpson's diversity ($D_{1/D}$)
* Dominance: Absolute ($N_{max}$), Berger-Parker index, $N_{max}/N$
* Productivity in numbers of individuals and in amount of cell quotas
* Residence time:
	* Ecosystem, $(length * width)/flowrate$
	* Individual, avg time spent in the system before dying or emigrating
	* Resource, avg time spent in the system before washing out or being consumed
	* Tracer, avg time spent in the system before washing out
* Individual residence time distribution (RTD)
* Resource particle RTD
* Tracer particle RTD

These data are stored in file as R-formatted data.frames. These files can be directly imported into an R or Python environment.

**Initialization**  
The model initiates with a random set of values for state-variables, 100 to 10,000 randomly drawn individuals from a theoretical log-series metacommunity. These values are saved, so that a simplex model could be programmed to replicate an analysis.

**Input data**  
simplex models require no input data, but it might be cool to use an api to grab environmental data or other 
data from a website to parameterize a simplex model.

**Submodels & Equations**

*Cell quota model of Droop*. In simplex models, individuals grow according to their amounts of endogenous resources (cell quota).

Droop (1968, 1983) gave a relationship between specific growth rate ($\mu$) and cell quota ($Q$): $$\mu = \mu'_{m} (1 - k_{q}/Q)$$
where $k_{q}$ is the minimum cell quota needed for life, also referred to as the subsistence quota. $\mu'$ is the 

*Maintenance cost of Pirt*. Pirt (1965) states "The variation, with growth rate, of the yield of organism from the substrate used as energy source is attributed to consumption of energy at a constant rate for cell maintenance." He derived a relationship between the growth yield (biomass), the growth rate, and metabolic maintenance. simplex models use Pirt's concept of a constant maintenance requirement. simplex also draws from Pirt's simple relation for substrate use: $$use(total) = use(maintenance) + use(growth)$$

Note that respiration and activity without growth is not accounted for.

*Log-series metacommunity*. simplex models draw immigrating individuals from a theoretical log-series distribution.
Hubbell (2001) states that the regional community (i.e., metacommmunity) often has a log-series species 
abundance distribution. The probability density function (pdf) of the log-series is: $$f(k) = -1/ln(1 - p) * p^{k}/k$$


**Notes on simplex source code**  
\setlength\parindent{36pt}
simplex models operate primarily on lists in a programmatic way, e.g., quickly sorting lists, and removing and returning an element from lists with very little overhead. Likewise, simplex models generate and hold a lot of information about all the particles and elements in the system, which can become a computationally intensive task. To this end, simplex modeling coded is written in Python, an easy to read high-level programming language that has many scientific, plotting, and animation libraries. Python gives greater control over the operating system than data analysis languages (e.g. R, Matlab) that can be comparatively slow at purely computational tasks and can greatly limit the amount of memory held in any data object, and even fail to import large amounts of data. Python can also obtain C-like speeds when implementing certain software, e.g., Cython, Sage.

The output of simplex is a broad array of information held in seven .csv files. The most important of these is SimData.csv, and is intended to be analyzed in the freely available R (https://www.r-project.org/) and RStudio (https://www.rstudio.com/) environments. The R statistical computing language is well-suited to the analysis of simplex output and contains many packages for multivariate analysis and higher-order statistical analysis that Python is only beginning to accumulate. Consequently, I provide R source code in .R files and .Rmd (RMarkdown) files, complete with basic and advanced statistical analyses for analyzing diversity, regression models, ordination, variance partitioning, and for generating pdf documents (via Knitr) that integrate prose, code, and figures for manuscripts. 

The reader can view example R-based analysis files that users can use to examine simplex's simulated data: https://github.com/LennonLab/simplex/tree/master/results/analyses.

\setlength\parindent{0pt}

**Speed & Memory**  

\setlength\parindent{36pt}
Simplex models do not complete until the time series of total abundance values reaches a state of mean reversion (i.e., stationarity). Because simplex models can range from quickly flowing systems with high disturbance to barely flowing and highly stable but depleted systems, simulations can potentially take several minutes or more to complete. Likewise, the ability to simulate many complex scenarios also allows for very large total abundances, the values of which cannot be predicted *a priori* and can even potentially outstrip a computer's memory. 

I ran simplex on a Mid 2010 MacBook Pro (OS X 10.9.5) with a 2.4 GHz Intel Core 2 Duo processor and 4GB of Memory. This system probably represents a below average capacity for modern personal computers, which for this study, was desirable as simplex should be able to be ran on both personal computers and high capacity remote servers. I present results for time to completion and required memory in the Results.

\setlength\parindent{0pt}

**Animations**  
Simplex can generate animations of its models and store them in various image file formats. It does this using the matplotlib animation libary. At the moment, choosing whether to animate or not animate a model is done by commenting out lines of code in model.py. However, this feature will be developed more highly for future versions.


**RESULTS**

**Unit tests**  
simplex passed all units tests for 15 diversity indices, ensuring that each index returns either the correct calculated value, or 'NaN' if given any values that cannot be used (e.g., negative numbers, string characters, empty lists).

**Speed & Memory**  
The following are results from 100 randomly assembled models. On average, simplex models required 96.60 +- 32.70 seconds to run, had an average total abundance (*N*) of 31,367 +- 879, and required 161 +- 4.3 megabytes of memory. The longest any model took to complete was 49 minutes. This particularly demanding model required 261MB of memory and also had generated the greatest total abundance (*N* = 68,808). The shortest any model took to complete was 5.8 seconds with a total abundance of 0 individuals. The least amount of memory any model required was 90MB. These and other analyses from the 100 randomly assembled models can be run using the SpeedMemory.Rmd file located in the results/TestResults directory.

**Products**  
Simplex generates six files as its output data. They are: 

*SimData.csv*. Formatted as an R data frame, where each row is a run of a randomly assembled model, and each column holds a piece of data about the system that was modeled (e.g., flow rate, total abundance, species richness, species turnover, rate of disturbance, etc.).  

*IndRTD.csv*, *ResRTD.csv*, *TracerRTD.csv*. Each line of these three files contains the amount of time that each individual, resource particle, or inert tracer spent in the system. Each line represents the same run of a randomly assembled model.

*RADs.csv*. Each line is a run of a randomly assembled model and contains the rank-abundance vector at the point when the model was stopped.

*Species.csv*. Each line is a run of a randomly assembled model and contains a vector of species labels corresponding to RADs.csv.

All output data were correctly formatted and placed within results/simulated_data/examples directory. Each of the six output data files was able to be imported into the RStudio environment using the Exploratory.Rmd Rmarkdown file provided in the "GitHub/simplex/results/analyses/Rmd/" path. The user can use the Exploratory.Rmd file to craft a .Rmd file for their own specific analyses.

**R analysis files**\setlength\parindent{36pt}

There are currently two R Markdown analysis files associated with simplex. These are located in the tools/analyses/Rmd directory and are designed to help users understand how to import simplex files and use R Markdown to analyze simplex data. The first file, `BasicAnalysis.Rmd`, focuses on descriptive statistics, multiple univariate analyses, and variance partitioning to reveal relationships among simplex variables. Running this file on example simplex data revealed that simplex produces well-known and/or intuitive relationships (Figs 1, 2, 3, 4).

```{r, results = 'hide', echo=F, message = F, warning = F}
rm(list=ls()) 
getwd()
setwd("~/GitHub/simplex") 

#install.packages("vegan") # Example of how an install can be done
require("vegan") 
require("car") 

# A table where each columns corresponds to a state variable or model output.  
simplex.dat <- read.csv("~/GitHub/simplex/results/simulated_data/examples/SimData.csv")

# Replacing 0's with 1's to allow log-transforms below
simplex.dat <- simplex.dat[complete.cases(simplex.dat), ]

simplex.dat$total.abundance[simplex.dat$total.abundance<=0] <- 0.00001
simplex.dat$N.max[simplex.dat$N.max<=0] <- 0.00001
simplex.dat$species.richness[simplex.dat$species.richness<=0] <- 0.00001
simplex.dat$resource.particles[simplex.dat$resource.particles<=0] <- 0.00001
simplex.dat$resource.concentration[simplex.dat$resource.concentration<=0] <- 0.00001
simplex.dat$Whittakers.turnover[simplex.dat$Whittakers.turnover<=0] <- 0.00001
simplex.dat$Jaccards.dissimilarity[simplex.dat$Jaccards.dissimilarity<=0] <- 0.00001

simplex.dat$total.abundance <- log(simplex.dat$total.abundance)
simplex.dat$N.max <- log(simplex.dat$N.max)
simplex.dat$Whittakers.turnover <- log(simplex.dat$Whittakers.turnover)
simplex.dat$Jaccards.dissimilarity <- log(simplex.dat$Jaccards.dissimilarity)
simplex.dat$species.richness <- log(simplex.dat$species.richness)
simplex.dat$resource.particles <- log(simplex.dat$resource.particles)
simplex.dat$resource.concentration <- log(simplex.dat$resource.concentration)
```

**Figure 1.** A scatter plot matrix revealing that the values of inputs representing physical variables were independent of one another. This is only a subset of the physical variables in simplex. Amplitude pertains to magnitude of temporal fluctuation in the inflow of propagules and/or resources. Barriers pertains to the number of barriers in the system, which varied between 1 and 4.

```{r, results='hide', echo=T, message=F, warning=F, fig.width=7, fig.height=6}
# Physical and metacommunity variables
png(filename="~/GitHub/simplex/results/figures/examples/PhysicalUnivariate.png")

phys.dat <- as.matrix(subset(simplex.dat, 
                      select = c(flow.rate,
                                 barriers,
                                 width,
                                 height,
                                 amplitude)))

scatterplotMatrix(phys.dat, main="Physical data", diagonal = "none")
dev.off()
```

**Figure 2.** A scatter plot matrix revealing that the values of resource-related variables. This is only a subset of resource variables in simplex. Number of resource particles, their concentration, resource richness, and resource diversity (a combination of resource richness and the variance in abundance among resources) are all highly and positively correlated.

```{r, results='hide', echo=T, message=F, warning=F, fig.width=7, fig.height=6}
# Resource variables
png(filename="~/GitHub/simplex/results/figures/examples/ResourceUnivariate.png")

res.dat <- as.matrix(subset(simplex.dat, 
                      select = c(resource.concentration,
                      shannons.resource.diversity,
                      resource.richness,
                      resource.particles)))

scatterplotMatrix(res.dat, main="Resource data", diagonal = "none")
dev.off()
```

**Figure 3.** A scatter plot matrix revealing that the values of physiological variables. This is only a subset of those in simplex. As of 29 October 2015, simplex does not include explicit stoichiometry. An intuitive life history trade-off arises in that increasing growth rate leads to increasing metabolic maintenance.

```{r, results='hide', echo=T, message=F, warning=F, fig.width=7, fig.height=6}
# Physiological variables
png(filename="~/GitHub/simplex/results/figures/examples/PhysioUnivariate.png")

physio.dat <- as.matrix(subset(simplex.dat, 
                      select = c(biomass.prod.N,
                      avg.per.capita.N.efficiency,
                      max.growth.rate,
                      max.met.maint)))

scatterplotMatrix(physio.dat, main="Physiological data", diagonal = "none")
dev.off()
```

**Figure 4.** A scatter plot matrix revealing that the values of diversity-related variables. This is only a subset of those in simplex. We see a strong postivie relationships between total numbers of individuals (total abundance) and total numbers of species (species richness) and between total abundance and the abundance of the most abundant species (Nmax). Each of these relationships is widely observed in ecological communities (Locey and Lennon 2015).

```{r, results='hide', echo=T, message=F, warning=F, fig.width=7, fig.height=6}
# Diversity-related variables
png(filename="~/GitHub/simplex/results/figures/examples/DiversityUnivariate.png")

diversity.dat <- as.matrix(subset(simplex.dat, 
                      select = c(total.abundance,
                      N.max,
                      species.richness,
                      simpson.e,
                      Whittakers.turnover,
                      Jaccards.dissimilarity)))

scatterplotMatrix(diversity.dat, main="Diversity data", diagonal = "none")
dev.off()
```


The second file, `Diversity.Rmd`, focuses on the analysis of taxonomic diversity and abundance to include fitting species abundance models using the R-based `vegan` module. Each file generates a knitted pdf (includes code, text, and figures), see Supplement. 

**Animations**   
I used the animation module in the matplotlib plotting library to generate avi formatted movies of simplex models. These can be viewed in any internet brower (https://github.com/LennonLab/simplex/tree/master/results/movies/examples). Currently, generating movies requires following three easy steps explained in model/model.py.

\setlength\parindent{0pt}

**DISCUSSION**

\setlength\parindent{36pt}

The simplex platform assembles individual-based models from random combinations of state-variables and processes, to explore a degrees of variation in ecological conditions and complexity. In this way, simplex allows ecologists to examine whether the influence of particular variables should be robust to a wide range of ecological and environmental conditions. In addition to its unique modeling aspects, simplex includes analysis code that shows users how to import and analyze simulated datasets, including variance partitioning and fitting of species abundance models.

Future developments of simplex will have trophic dynamics including predator-prey, symbionts and mutualists, and producers and consumers. Individuals will also be attributed with additional behaviors (e.g., search patterns), sensing (e.g., of resources), and learning (e.g., predator avoidance). Additionally, individuals will also be attributed with genomes. Initially, these genomes will be represented by sets of sequences of A's, C's, T's, and G's representing genes that can be mutated as they're passed on through clonal reproduction, or mutated and recombined in the case of sexual haploid or diploid reproduction. A precursor to this has already been developed in Locey (2013), the code for which is available here: https://github.com/klocey/SYMENT.

By taking advantage of modern computing and version control and social coding tools (e.g., Python, R/RStudio, Knitr, Git, GitHub) simplex is well-positioned to continue development as both a research and education/instructional tool for ecology. Future developments and released versions will include additional ecological dynamcis such as predator-prey, mutualism, and parasitism. Improvements to simplex will also include an explicit stoichiometry, where individual resource particles have ratios of Carbon, Nitrogen, and Phosphorus, as well as degrees of biocomplexity. For example, given the chemical structure of phosphate and the size of a resource particle, the particle can be assigned *x* units of phosphorus as well as the biocomplexity value of phosphate (estimated as a form of Shannon's entropy). 

Likewise, improvements and future versions of simplex will provide increasing numbers of files for statistical analysis of simplex output. These files will, in a sense, be able to be used as a complete analysis of some particular aspect of simplex output (e.g., demographic, resource-related, spatial, multivariate, ecological model testing). These and all simplex files are freely available and open source, allowing users to modify simplex source code to suit their particular needs.

\setlength\parindent{0pt}

**ACKNOWLEDGMENTS**

I thank Jay T. Lennon, Mario Muscarella, Xiao Xiao, and Ed Hall for fruitful discussions. I thank Dr. Daniel V. Schroeder for his web-based fluid dynamic simulator, the python-based code for which serves as the core method in simplex for simulating fluid flow.

**REFERENCES**

1. Black AJ, McKane AJ (2012) Stochastic formulation of ecological models and their applications. Trends in Ecology and Evolution. 27:337-345.

2. DeAngelis DL, Gross LJ (1992) Individual-based Models and Approaches in Ecology: Populations, Communities, and Ecosystems. Chapman & Hall. New York.

3. Droop MR (1983) Twenty-five years of algal growth kinetics, a personal view. Botanica Marina. 26: 99-112.

4. Gotelli NJ, Entsminger GL (2001) Swap and fill algorithms in null model analysis: rethinking the knight’s tour. Oecologia. 129:281–291

5. Grimm V, Berger U, Bastiansen F, Eliassen S, Ginot V, Giske J, Goss-Custard J, Grand T, Heinz SK, Huse G, Huth A, Jepsen JU, Jørgensen C, Mooij WM, Mu¨ller B, Pe’er G, Pioub C, Railsback SF, Robbins AM, Robbins MM, Rossmanith E, Ru¨ger N, Strand E, Souissim S, Stillmane RA, Vabøg R, Visser U, DeAngelis DL (2006) A standard protocol for describing individual-based and agent-based models. Ecological Modeling. 198: 115-126.

6. Grimm V, Railsback SF (2005) Individual-based Modeling and Ecology. Princeton University Press, Princeton.

7. Hartl DL, Clark AG (1997) Principle of Population Genetics. Sinauer Associates, Inc. Publishers, Sunderland, MA.

8. Hubbell SP (2001) The unified neutral theory of biodiversity and biogeography. Princeton University Press, Princeton, NJ.

9. Leadbeater BSC (2006) The 'Droop equation' - Michael Droop and the legacy of the 'Cell-quota model' of phytoplankton growth. Protist. 157: 345-358.

10. Locey K (2013) Synthesized ecological molecular neutral theory. figshare. http://dx.doi.org/10.6084/m9.figshare.640105.

11. Locey KJ, Lennon JT. (2015) Scaling laws predict global microbial diversity. PeerJ PrePrints 3:e1808 https://dx.doi.org/10.7287/peerj.preprints.1451v1

12. Matthews RB, Gilbert NG, Roach A, Polhill JG, Gotts NM (2007) Agent-based land-use models: a review of applications. Landscape Ecology. 22: 1447-1459.

13. Pirt SJ (1965) The Maintenance Energy of Bacteria in Growing Cultures. Proceedings of the Royal Society of London. Series B, Biological Sciences, 163:224-231.

14. Rosindell J, Harmon LJ, Etienne RS (2015) Unifying ecology and macroevolution with individual-based theory. Ecology Letters. 18: 472-482.

15. Succi S (2001) The Lattice Boltzmann Equation: For Fluid Dynamics and Beyond. Oxford University Press. New York.